src/cluster.c

```diff
@@ -852,176 +689,312 @@ static dict *parse_cluster_slots(valkeyClusterContext *cc, valkeyReply *reply) {
     return NULL;
 }
 
-/**
- * Parse the "cluster nodes" command reply to nodes dict.
- */
-static dict *parse_cluster_nodes(valkeyClusterContext *cc, valkeyReply *reply) {
-    int ret;
-    dict *nodes = NULL;
-    dict *nodes_name = NULL;
-    valkeyClusterNode *master, *slave;
-    cluster_slot *slot;
-    char *pos, *start, *end, *line_start, *line_end;
-    char *role;
-    int role_len;
-    int slot_start, slot_end, slot_ranges_found = 0;
-    sds *part = NULL, *slot_start_end = NULL;
-    int count_part = 0, count_slot_start_end = 0;
-    int k;
-    int len;
+/* Store a parsed replica node in given dict using the primary_id as key.
+ * Additional replicas for a primary are stored within the first replica. */
+static int store_replica_node(dict *replicas, char *primary_id, valkeyClusterNode *node) {
+    sds key = sdsnew(primary_id);
+    if (key == NULL)
+        return VALKEY_ERR;
 
-    if (reply->type != VALKEY_REPLY_STRING) {
-        valkeyClusterSetError(cc, VALKEY_ERR_OTHER, "Unexpected reply type");
-        goto error;
+    dictEntry *de = dictFind(replicas, key);
+    if (de == NULL) {
+        if (dictAdd(replicas, key, node) != DICT_OK) {
+            sdsfree(key);
+            return VALKEY_ERR;
+        }
+        return VALKEY_OK;
     }
 
-    nodes = dictCreate(&clusterNodesDictType, NULL);
-    if (nodes == NULL) {
-        goto oom;
+    /* Store replica node in the existing replica node. */
+    sdsfree(key);
+    valkeyClusterNode *n = dictGetEntryVal(de);
+    if (n->slaves == NULL) {
+        n->slaves = listCreate();
+        if (n->slaves == NULL)
+            return VALKEY_ERR;
+        n->slaves->free = listClusterNodeDestructor;
     }
+    if (listAddNodeTail(n->slaves, node) == NULL)
+        return VALKEY_ERR;
```

Comment:
This seems very weird.

Should the replicas be added under `n->slaves` on the primary?

From what I can read here, `n` is a replica and the other replicas are stored under `n->slaves`.